scenarios/command/c/command_client/main.c

```diff
@@ -0,0 +1,205 @@
+#define PUB_TOPIC "vehicles/vehicle03/command/unlock/request"
+#define COMMAND_CONTENT_TYPE "application/protobuf"
+#define COMMAND_TIMEOUT_SEC 10
+
+#define QOS 1
+#define MQTT_VERSION MQTT_PROTOCOL_V5
+
+#define UUID_LENGTH 37
+
+#define CONTINUE_IF_ERROR(rc)                                   \
+  do                                                            \
+  {                                                             \
+    if (rc != MOSQ_ERR_SUCCESS)                                 \
+    {                                                           \
+      printf("Error publishing: %s\n", mosquitto_strerror(rc)); \
+      mosquitto_property_free_all(&proplist);                   \
+      continue;                                                 \
+    }                                                           \
+  } while (0)
+
+static uuid_t pending_correlation_id;
+static time_t last_command_time;
+
+// Custom callback for when a message is received.
+// prints the message information from the command response and validates that the correlation data
+// matches.
+void handle_message(
+    struct mosquitto* mosq,
+    const struct mosquitto_message* message,
+    const mosquitto_property* props)
+{
+  printf(
+      "on_message: Topic: %s; QOS: %d; protobuf Payload: %s\n",
+      message->topic,
+      message->qos,
+      (char*)message->payload);
+
+  void* correlation_data;
+  uint16_t correlation_data_len;
+  if (mosquitto_property_read_binary(
+          props, MQTT_PROP_CORRELATION_DATA, &correlation_data, &correlation_data_len, false)
+      == NULL)
+  {
+    printf("Error reading correlation data\n");
+    return;
+  }
+
+  char readable_correlation_data[UUID_LENGTH];
+  uuid_unparse(correlation_data, readable_correlation_data);
+  printf("\tcorr_data: %s\n", readable_correlation_data);
+  if (uuid_compare(pending_correlation_id, correlation_data) != 0)
+  {
+    uuid_unparse(pending_correlation_id, readable_correlation_data);
+    printf("\t[ERROR] Correlation data does not match, expected: %s\n", readable_correlation_data);
+  }
+  else
+  {
+    uuid_clear(pending_correlation_id);
+  }
+
+  free(correlation_data);
+}
+
+/* Callback called when the client receives a CONNACK message from the broker and we want to
+ * subscribe on connect. */
+void on_connect_with_subscribe(
+    struct mosquitto* mosq,
+    void* obj,
+    int reason_code,
+    int flags,
+    const mosquitto_property* props)
+{
+  on_connect(mosq, obj, reason_code, flags, props);
+
+  int result;
+  //   mqtt_client_obj* client_obj = (mqtt_client_obj*)obj;
+  //   char sub_topic[strlen(client_obj->client_id) + 33];
+  //     sprintf(sub_topic, "vehicles/%s/command/unlock/response", client_obj->client_id);
+
+  /* Making subscriptions in the on_connect() callback means that if the
+   * connection drops and is automatically resumed by the client, then the
+   * subscriptions will be recreated when the client reconnects. */
```

Comment:
on a clean session?

Reply:
True - this is something we might need to reconsider if/when we add proper support for clean session = false.